pkg/virt-controller/watch/volume-migration/volume-migration.go

```diff
@@ -0,0 +1,334 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright The KubeVirt Authors
+ *
+ */
+
+package volumemigration
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+
+	k8sv1 "k8s.io/api/core/v1"
+	virtv1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+	"kubevirt.io/kubevirt/pkg/controller"
+	storagetypes "kubevirt.io/kubevirt/pkg/storage/types"
+)
+
+const invalidUpdateErrMsg = "The volume can only be reverted to the previous version during the update"
+
+// invalidVols includes the invalid volumes for the volume migration
+type invalidVols struct {
+	hotplugged []string
+	fs         []string
+	shareable  []string
+	luns       []string
+}
+
+func (vols *invalidVols) errorMessage() error {
+	var s strings.Builder
+	if len(vols.hotplugged) < 1 && len(vols.fs) < 1 &&
+		len(vols.shareable) < 1 && len(vols.luns) < 1 {
+		return nil
+	}
+	s.WriteString("invalid volumes to update with migration:")
+	if len(vols.hotplugged) > 0 {
+		s.WriteString(fmt.Sprintf(" hotplugged: %v", vols.hotplugged))
+	}
+	if len(vols.fs) > 0 {
+		s.WriteString(fmt.Sprintf(" filesystems: %v", vols.fs))
+	}
+	if len(vols.shareable) > 0 {
+		s.WriteString(fmt.Sprintf(" shareable: %v", vols.shareable))
+	}
+	if len(vols.luns) > 0 {
+		s.WriteString(fmt.Sprintf(" luns: %v", vols.luns))
+	}
+
+	return fmt.Errorf(s.String())
+}
+
+func updatedVolumesMapping(vmi *virtv1.VirtualMachineInstance, vm *virtv1.VirtualMachine) map[string]string {
+	updateVols := make(map[string]string)
+	vmVols := make(map[string]string)
+	// New volumes
+	for _, v := range vm.Spec.Template.Spec.Volumes {
+		if name := storagetypes.PVCNameFromVirtVolume(&v); name != "" {
+			vmVols[v.Name] = name
+		}
+	}
+	// Old volumes
+	for _, v := range vmi.Spec.Volumes {
+		name := storagetypes.PVCNameFromVirtVolume(&v)
```

Comment:
How about:

```golang
if name == ""{
    continue
}
```

Thinking of the case where volume changes from containerdisk to PVC